### PR TITLE
AesKey: replace o.bouncycastle.c.p.KeyParameter

### PR DESCRIPTION
**Pull request by msgilligan**
We were using KeyParameter as a wrapper for a byte array containing an AES Key and passing it around our API. This introduced an unnecessary API dependency on Bouncy Castle. This PR adds an AesKey class for the same purpose to replace it.

This is a breaking change, but one that should be easy to accommodate.

### CODE REVIEW

**Comment by msgilligan** [comment]
@schildbach If this approach looks good, I'll make another pass and update all the JavaDoc.

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ *  to {@code AesKey}, you can use {@link #ofKeyParameter(KeyParameter)} or {@link #toKeyParameter()}
+ *  // TODO: Update JavaDoc
+ */
+public class AesKey {
+    private final byte[]  key;
+
+    public AesKey(
+            byte[]  key)
+    {
+        this.key = key;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/crypto/AesKey.java:38]: Note that the replaced class is taking ownership of the key data (copying) here.

Nitpick: the formatting could be improved around this line.

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bitcoinj.crypto;
+
+import org.bouncycastle.crypto.params.KeyParameter;
+
+/**
+ *  Wrapper for a byte array containing an AES Key. This is a replacement for Bouncy Castle's{@link KeyParameter} which
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/crypto/AesKey.java:22]: Missing space.

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+public class AesKey {
+    private final byte[]  key;
+
+    public AesKey(
+            byte[]  key)
+    {
+        this.key = key;
+    }
+
+    public byte[] getKey()
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/crypto/AesKey.java:41]: Since the class name already has "key in its name, this could be named just "get()".

**Comment by schildbach** [comment]
I just checked with my app and can testify that while this change breaks some code, it looks easy to fix and it indeed gets rid of BouncyCastle imports apart from my own "Crypto" class.

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java (hunk 3 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/crypto/AesKey.java:41]: I wanted to keep the same accessor name as the replaced class. But we don't have to, we could also provide a deprecated `getKey()` method along with one with a new name.

We could also name the accessor `bytes()`.

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java (hunk 1 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/crypto/AesKey.java:38]: > Note that the replaced class is taking ownership of the key data (copying) here.
> 

My mistake. `byte[]` is mutable and should be defensively copied.

**Comment by msgilligan** [comment]
Rebased to `master`

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java (hunk 3 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/crypto/AesKey.java:41]: @schildbach do you agree with a deprecated `getKey()` and a new accessor name of `bytes()`?

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java (hunk 1 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/crypto/AesKey.java:38]: Fixed.

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java (hunk 3 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/crypto/AesKey.java:41]: I'm going to assume that is ok with you and make the change. If you don't like it, let me know in the next review.

**Comment by msgilligan** [comment]
@schildbach All issue addressed, force-pushed, and ready-to-review/merge.

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ */
+public class AesKey {
+    private final byte[] bytes;
+
+    /**
+     * Wrapper for a {@code byte[]} containing an AES Key
+     * @param keyBytes implementation-dependent AES Key bytes
+     */
+    public AesKey(byte[] keyBytes)
+    {
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/crypto/AesKey.java:38]: opening brace at wrong position? (3 occurences)

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java (hunk 3 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/crypto/AesKey.java:58]: Yes, that's fine as well. Actually I had this thought, too.

**Review by schildbach — COMMENTED** [review]
Looks great!

File: core/src/main/java/org/bitcoinj/crypto/AesKey.java (hunk 4 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/crypto/AesKey.java:38]: I'll fix it.

Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/crypto/AesKey.java]: Fixed.

**Comment by msgilligan** [comment]
@schildbach OK, 3 opening-brace issues fixed.  (I'm gonna blame Bouncy Castle and IntelliJ for that 🤷‍♂️)

**Comment by schildbach** [comment]
Merged!

**Pull request closed without merging**